tsfile/src/main/java/org/apache/iotdb/tsfile/file/metadata/TsFileMetaData.java

```diff
@@ -1,306 +1,387 @@
+    return measurementSchema;
+  }
+
+  /**
+   * use the given outputStream to serialize.
+   *
+   * @param outputStream -output stream to determine byte length
+   * @return -byte length
+   */
+  public int serializeTo(OutputStream outputStream) throws IOException {
+    int byteLen = 0;
+
+    byteLen += ReadWriteIOUtils.write(deviceIndexMap.size(), outputStream);
+    for (Map.Entry<String, TsDeviceMetadataIndex> entry : deviceIndexMap.entrySet()) {
+      byteLen += ReadWriteIOUtils.write(entry.getKey(), outputStream);
+      byteLen += entry.getValue().serializeTo(outputStream);
+    }
+
+    byteLen += ReadWriteIOUtils.write(measurementSchema.size(), outputStream);
+    for (Map.Entry<String, MeasurementSchema> entry : measurementSchema.entrySet()) {
+      byteLen += ReadWriteIOUtils.write(entry.getKey(), outputStream);
+      byteLen += entry.getValue().serializeTo(outputStream);
+    }
+
+    byteLen += ReadWriteIOUtils.writeIsNull(createdBy, outputStream);
+    if (createdBy != null) {
+      byteLen += ReadWriteIOUtils.write(createdBy, outputStream);
+    }
+
+    byteLen += ReadWriteIOUtils.write(totalChunkNum, outputStream);
+    byteLen += ReadWriteIOUtils.write(invalidChunkNum, outputStream);
+
+    return byteLen;
+  }
+
+  /**
+   * use the given outputStream to serialize bloom filter.
+   *
+   * @param outputStream -output stream to determine byte length
+   * @return -byte length
+   */
+  public int serializeBloomFilter(OutputStream outputStream,
+      List<ChunkGroupMetaData> chunkGroupMetaDataList)
+      throws IOException {
+    int byteLen = 0;
+    BloomFilter filter = buildBloomFilter(chunkGroupMetaDataList);
+
+    byte[] bytes = filter.serialize();
+    byteLen += ReadWriteIOUtils.write(bytes.length, outputStream);
+    outputStream.write(bytes);
+    byteLen += bytes.length;
+    byteLen += ReadWriteIOUtils.write(filter.getSize(), outputStream);
+    byteLen += ReadWriteIOUtils.write(filter.getHashFunctionSize(), outputStream);
+    return byteLen;
+  }
+
+  /**
+   * get all path in this tsfile
+   *
+   * @return all path in set
+   */
+  private List<String> getAllPath(List<ChunkGroupMetaData> chunkGroupMetaDataList) {
+    List<String> res = new ArrayList<>();
+    for (ChunkGroupMetaData chunkGroupMetaData : chunkGroupMetaDataList) {
+      String deviceId = chunkGroupMetaData.getDeviceID();
+      for (ChunkMetaData chunkMetaData : chunkGroupMetaData.getChunkMetaDataList()) {
+        res.add(deviceId + PATH_SEPARATOR + chunkMetaData.getMeasurementUid());
+      }
+    }
+
+    return res;
+  }
+
+  /**
+   * build bloom filter
+   *
+   * @return bloom filter
+   */
+  private BloomFilter buildBloomFilter(List<ChunkGroupMetaData> chunkGroupMetaDataList) {
+    BloomFilter bloomFilter = BloomFilter
+        .getEmptyBloomFilter(TSFileDescriptor.getInstance().getConfig().getBloomFilterErrorRate(),
+            chunkGroupMetaDataList.size());

```

Comment:
I have a question: chunkGroupMetaDataList.size() doesn't seem to equal the total number of string added to the bloom filter. The total number of strings added is the size of the returned list from `getAllPath` function, right?

Reply:
Yes, I think this is a bug. So I have to build list of path and the add it to bloom filter. Only by this I can get the size of the bloom filter